cmake/FindCharm.cmake

```diff
@@ -78,62 +211,145 @@ endif()
 set(CHARM_VERSION
   "${CHARM_VERSION_MAJOR}.${CHARM_VERSION_MINOR}.${CHARM_VERSION_PATCH}")
 
-find_library(CHARM_LIBCK
-    NAMES ck
-    PATH_SUFFIXES lib
-    HINTS ${CHARM_ROOT} ENV CHARM_ROOT
-)
+# Filter the compiler and linker flags:
+# - Remove the macOS deployment target so the CMake setting is used.
+list(FILTER CHARM_CXX_FLAGS EXCLUDE REGEX "^-mmacosx-version-min=")
+list(FILTER CHARM_LDXX_FLAGS EXCLUDE REGEX "^-mmacosx-version-min=")
+# - Remove the C++ standard flag so the CMake setting is used.
+list(FILTER CHARM_CXX_FLAGS EXCLUDE REGEX "^-std=")
+# - Remove the standard library flag so the CMake setting is used.
+list(FILTER CHARM_CXX_FLAGS EXCLUDE REGEX "^-stdlib=")
+list(FILTER CHARM_LDXX_FLAGS EXCLUDE REGEX "^-stdlib=")
+# - Remove the include directory so we can set it properly with CMake. That's
+#   better because it is declared as "SYSTEM", which silences warnings from
+#   those headers.
+list(FILTER CHARM_CXX_FLAGS EXCLUDE REGEX "^-I${CHARM_INCLUDE_DIR}$")
+# - Remove the link directory and find the libraries on the system so we can
+#   configure them properly with CMake. That's more robust when switching
+#   between static and shared libs builds. Instead of parsing the list of libs
+#   below we could just hard-code lib names, but that may break with future
+#   Charm versions and could be annoying to debug. However, it would give us
+#   more control over which libs to link. Ideally, Charm++ would provide
+#   exported targets with these libs.
+list(FILTER CHARM_LDXX_FLAGS EXCLUDE REGEX "^-L${CHARM_LIBRARIES}$")
+set(CHARM_LIB_NAMES ${CHARM_LDXX_FLAGS})
+list(FILTER CHARM_LIB_NAMES INCLUDE REGEX "^-l.+")
+list(FILTER CHARM_LDXX_FLAGS EXCLUDE REGEX "^-l.+")
+list(TRANSFORM CHARM_LIB_NAMES REPLACE "^-l" "")
+list(REMOVE_DUPLICATES CHARM_LIB_NAMES)
```

Comment:
Not sure why I added this line, probably because I thought it silly to link some libs multiple times. Deleted the line in a fixup. Please confirm this is the right thing to do in your opinion.

Reply:
Yes, this looks right.  (I've had to point this out to compiler authors, so don't feel too bad about missing it.)